### ISSUE TITLE
GridAggregator - Unwanted border over image with patch_overlap & floats not welcome

### ISSUE THREAD

**Issue by ibro45** [fepegar/torchio#96]
**Describe the bug**
There are two bugs related to GridAggregator:

1. If `add_batch` is given a normalized, float, tensor (e.g. between -1 and 1), the final output from `get_output_tensor()` will be full of zeros (black). It works once I denormalize my tensors to `[0,255]` or `[-1024,3072]`.

2. `patch_overlap` produces border of the size of the overlap. You can see in [this gif](https://imgflip.com/gif/3ruqtf) from results for three different `patch_overlap`'s, `(0,0,0)`, `(4,4,4)` and `(16,16,16)`. In this example the border is grayish since the images shown were in range `-1024,3072` and the border produced apparently has values `0`.

**To Reproduce**
1. Use a 3D input with `GridSample` and `GridAggregator` as shown in the Readme
2. Make sure that just before the `add_batch` method the input tensors values are shifted so that 0 is somewhere inside of the range (e.g `[-128,128] or [-1024,3072]`). This is just so that you can clearly see the border, if you were outputting segmentation masks, it won't be obvious since they're mostly black anyways.
3. Run it with different `patch_overlaps` and observe the border

**Expected behavior**
I assume that zeros are assigned in border with the size of the overlap since no overlap is happening on the edges of the image. I would expect that the border does not appear, but rather that the predicted values are shown even if they had no overlap. 

**Output**
Images from the gif, shown separately [https://imgur.com/a/Rk0fMd4](https://imgur.com/a/Rk0fMd4)
Example similar to what I did when I observed this issue:
```
patch_size = (128, 128, 128)
patch_overlap = (16, 16, 16)
batch_size = 1

grid_sampler = torchio.inference.GridSampler(
    data,
    patch_size=patch_size,
    patch_overlap=patch_overlap,
)
patch_loader = torch.utils.data.DataLoader(grid_sampler, batch_size=batch_size)
aggregator = torchio.inference.GridAggregator(
    data,
    patch_overlap=patch_overlap,
)

for patches_batch in patch_loader:
    input_tensor = patches_batch['image'].to(device)
    locations = patches_batch['location']
    out = model.forward(input_tensor)
    # assume output is in [0,1], let's make it [-128,127]
    out = out * 255 - 128
    aggregator.add_batch(out, locations)

output = aggregator.get_output_tensor()
plt.imsave('output.png', output.numpy()[300], cmap='gray')

**Comment by fepegar** [comment]
Hi @ibro45,

Thanks for such a detailed explanation.

#### Unwanted border over image with `patch_overlap`
The border issue is expected. It happens because only the central part of the inferred patch is kept.
See [NiftyNet docs](https://niftynet.readthedocs.io/en/dev/window_sizes.html) for some info about this.
The best you can do is pad the input image with `patch_size` voxels on each side and crop it at the end.

Despite this being expected, I agree it's not well documented. I'll work on this.

#### Floats not welcome
I haven't been able to reproduce this. Can you provide a self-contained example?

Are you sure it's not a data type issue? In your code, you're saving the image in PNG, which doesn't support floats, as far as I know. Can you print the min and max values in `output`? These classes shouldn't be messing with the voxel values.

**Comment by ibro45** [comment]
Thanks for quick response @fepegar! 

#### Unwanted border over image with `patch_overlap`

> The border issue is expected. It happens because only the central part of the inferred patch is kept.
See NiftyNet docs for some info about this.

I wasn't aware that that is the expected behaviour, I guess I'll just pad my inputs in that case. 

> Despite this being expected, I agree it's not well documented. I'll work on this.

A suggestion regarding the documentation, maybe it would be easier to use something like [readthedocs.org](https://readthedocs.org/), it would allow adding more details while maintaining readability.

#### Floats not welcome
What's happening in my case is that the output of the aggregator ends up being in type `int` even though the aggregated patches are of type `float`, as you can see [here](https://imgur.com/a/dYP4dnJ). The reason that happens is that the argument `data` was given data of type `int`, and the aggregator will cast the batches to that type even if the batches themselves are  in `float` (e.g. normalized). Here's example code for more intuition:

```
patch_size = (128, 128, 128)
patch_overlap = (16, 16, 16)
batch_size = 1

# data is in type "int", in range [-1024, 3072]
data = np.load('LIDC-IDRI-0095_1.3.6.1.4.1.14519.5.2.1.6279.6001.137773550852881583165286615668.npy')

grid_sampler = torchio.inference.GridSampler(
    data,
    patch_size=patch_size,
    patch_overlap=patch_overlap,
)
patch_loader = torch.utils.data.DataLoader(grid_sampler, batch_size=batch_size)
aggregator = torchio.inference.GridAggregator(
    data, # when aggregating, it will cast the batches to the type of the data, which is "int" in this case
    patch_overlap=patch_overlap,
)

for patches_batch in patch_loader:
    input_tensor = patches_batch['image'].to(device)
    input_tensor = normalize(input_tensor) # here we're normalizing it to [-1,1]
    locations = patches_batch['location']
    with torch.no_grad()
        out = model.forward(input_tensor)
    # the given "out" is float, but it will cast it to "int" because the "data" was in that format
    aggregator.add_batch(out, locations)

output = aggregator.get_output_tensor()
``` 

The problem is easily fixed by changing the `data` to type `float`.
```
data = np.load('LIDC-IDRI-0095_1.3.6.1.4.1.14519.5.2.1.6279.6001.137773550852881583165286615668.npy')
data = data.astype(float)
```
I guess it is not then an issue per se, but having it more clear in the documentation would probably help here too. And yes, thanks a lot for this library, really helps for faster development and with having less clutter in code when working with 3D images.

**Comment by fepegar** [comment]
> I wasn't aware that that is the expected behaviour, I guess I'll just pad my inputs in that case.

I wouldn't have been either, as this is not well documented.

> A suggestion regarding the documentation, maybe it would be easier to use something like readthedocs.org, it would allow adding more details while maintaining readability.

Thanks, I've been working on it recently! See a sample [here](https://torchio.readthedocs.io/data/queue.html). Unfortunately, RTD builds [have been failing](https://github.com/readthedocs/readthedocs.org/issues/6727) recently.

> What's happening in my case is that the output of the aggregator ends up being in type int even though the aggregated patches are of type float.

I think you're right and it makes sense that the data type of the output depends on the patches, not on the input image. I'll modify the behavior.

> data = np.load('LIDC-IDRI-0095_1.3.6.1.4.1.14519.5.2.1.6279.6001.137773550852881583165286615668.npy')

FYI I recently added support to read DICOM files, it might be useful for you.

> And yes, thanks a lot for this library, really helps for faster development and with having less clutter in code when working with 3D images.

Thanks for the kind words :) Hopefully, a paper will be published on Arxiv tonight!

**Comment by ibro45** [comment]
> Thanks, I've been working on it recently! See a sample here. Unfortunately, RTD builds have been failing recently.

Looking forward to it!

> FYI I recently added support to read DICOM files, it might be useful for you.

Good to know, I see you're familiar with the LIDC dataset :D 

> Thanks for the kind words :) Hopefully, a paper will be published on Arxiv tonight!

Cool, I will check it out and make sure to cite it in our work!

> I think you're right and it makes sense that the data type of the output depends on the patches, not on the input image. I'll modify the behavior.

Makes it more intuitive indeed, I've just made a pull request because a small mistake sneaked in.

**Comment by fepegar** [comment]
> I see you're familiar with the LIDC dataset :D

Not really! But that filename looked a lot like DICOM :)

Thanks again for the encouragement and for that fix.
